### ISSUE TITLE
Add back 8-bit palette based rendering mode

### ISSUE THREAD

**Issue by ihhub** [libsdl-org/SDL#6899]
SDL 1 has of 8-bit palette based video mode support. This mode is extremely helpful for video games which handle 8-bit graphics. We are working on **[fheroes2](https://github.com/ihhub/fheroes2)** project which by default uses 8-bit graphics. This is a game engine recreation of a 25-years game old game so we cannot change graphics.

SDL 2 somehow does not support this video mode which leads to about 20% (depending on frame size) performance drop in rendering as we have to convert 8-bit image into 32-bit image before rendering a frame. It would be awesome if SDL 2 starts supporting this mode at least for specific platforms. `SDL_PIXELFORMAT_INDEX8` format is not detected from the list of formats.

**Comment by bradallred** [comment]
PR #6192 might interest you

**Comment by slime73** [comment]
I think @icculus' comment https://github.com/libsdl-org/SDL/pull/6192#issuecomment-1325503480 is reasonable - using a palette in a shader is basically instant, probably much faster than any other solution while also being forward-compatible with new platforms and backends.

**Comment by ihhub** [comment]
Hi @slime73 , are you suggesting that we need to use DirectX / X11 / OpenGL code in combination with SDL to have palette based images support?

**Comment by slime73** [comment]
I'm saying using the GPU is probably going to be by far the most efficient method since you're concerned about performance. There's a new graphics API abstraction planned for SDL3 so you wouldn't have to write platform-specific DirectX/OpenGL/etc. code.

**Comment by ihhub** [comment]
Hi @slime73 , all the frame generation is done by our game engine, we use SDL only for rendering purposes when a frame is fully prepared (if we talk about rendering itself). SDL 3 might take time to develop and also to properly test before it can be fully marked as a stable release so for now our team is stuck with SDL 2. Hence we need to find a solution to improve the performance with existing libraries :) To be more specific our game engine generates roughly 60 FPS on some low-end machines while the real number should be about 125 FPS. With 20% boost this number will be roughly 70 FPS which is a might more reasonable performance. Translate these number into Raspberry Pi which usually has 30 FPS.

Alternatively, if you have time and willingness you can take a look into [screen.cpp](https://github.com/ihhub/fheroes2/blob/master/src/engine/screen.cpp) file, class **class RenderEngine** starting from line 738. Maybe we use SDL in a wrong way.

**Comment by ccawley2011** [comment]
There's still a use for having it in the render API since it can be used on platforms that don't support shaders, but do have hardware support for CLUT8 palettes (such as the PSP or PS2), as well as allowing it to transparently handle falling back on software conversion for platforms that don't support either of those things. It'll also be simpler for the end user to not have to create custom shaders for palette rendering.

**Comment by slime73** [comment]
> if you have time and willingness

I don't, but if you haven't already I definitely recommend using profiling tools to identify and optimize bottlenecks. The most expensive parts of a frame are often surprising compared to what you might expect.

> having it in the render API

Whether it fits in SDL_Render depends on the intended scope of the render API, I suppose. In the past it's been kept pretty minimal, which has also allowed it to be ported to a lot of different platforms and maintained without a huge amount of effort. Ryan's comment seems to suggest keeping it that way.

**Comment by ihhub** [comment]
Hi @slime73 , 8-bit to 32-bit conversion is the biggest bottleneck right now as our code is extremely optimized on all levels :)

**Comment by slime73** [comment]
Right, I'm suggesting there may be ways to optimize that in your code. And sometimes there are ways to optimize by reducing the amount of times it happens (sort of a bottom-up versus top-down approach to optimizing.) In many cases those two approaches can be combined, too.

**Comment by 1bsyl** [comment]
@ihhub quickly looking at your code...
there is the CopyImageToSurface, then there is a SDL_UpdateTexture with the surface->pixels.

I don't really got what is the "transform" path.  (palette to 32 bits ?)
but, there is also memcpy path that is definitively a bottleneck: 
so, whenever you use this memcpy path, you should instead  try to use SDL_UpdateTexture with the imageIn.image() ptr, with appropriate pitch if possible. .. you would avoid a memcpy per frame.

**Comment by ihhub** [comment]
Hi @1bsyl , transform is pointer to 32-bit array so:
for ( ; out != outEnd; ++out, ++in )
    *out = *( transform + *in );

is to convert from 8-bit to 32-bit image.

memcpy is used only for 8-but surfaces which are supported only by SDL 1. Yes, I agree with using pixels from a texture. I did small tests today and I haven't noticed the difference in performance in comparison if I remove conversion loop.

**Comment by 1bsyl** [comment]
to remove the conversion loop (and have something equivalent), you should use the palette texture PR. and you should definitively see something faster ? So I don't get it ..


what you can try also. is to use SDL2 internal function to perform the 8bit to 32 bits conversion because they may be faster (using assembly, duff loop, align pointer, not sure ..) . 
either: copy paste the relevant SDL blit code ..
( or use SDL_CreateRGBSurfaceFromFormat(imageIn.image, INDEX8), + set you palette. And  convert surface to 32bit using SDL_ConvertSurfaceFormat() )

**Comment by ihhub** [comment]
Hi @1bsyl , if was a typo: if --> unless :)

I will a try with textures and test on machines with much slower RAM as my machine is quite new so no surprise that memcpy doesn't make much difference.

**Comment by slouken** [comment]
This would be pretty reasonable to add, possibly with the new GPU API? I'm going to add this to the SDL 3.0 milestone for review.

**Comment by rofl0r** [comment]
wow bummer. i was just working on modifying my 32bit RGB code to use 8bit palette for drawing due to abysmall performance (16 fps on PSP @222 mhz). now it seems this will never be implemented for SDL2 and only eventually in SDL3 so i can use it maybe in 5 years when all distros provide SDL3, since i want to run my code on multiple platforms. guess i was lucky running into this issue while googling for "PSP fastest videomode" before i actually wrote all the code and would run into "Palettized textures are not supported".

**Comment by slouken** [comment]
Reading over the referenced PR, it looks like adding this will only add one additional entry point to the API, and is something we can add later, if bilinear filtering is solved at some point. For now I'm going to bump this to the 3.x milestone.

**Comment by ihhub** [comment]
It would be great if SDL3 would support 8-bit palette based images. It should give ~30% speed boost to fheroes2 project. We are waiting till it happens to migrate the project from SDL2🙂

**Comment by glebm** [comment]
We still have to use SDL1.2 in 2025 for ports of DevilutionX to older platforms that don't have accelerated graphics, which is unfortunate. It'd be nice to be able to target a single version of SDL.

**Comment by ccawley2011** [comment]
> We still have to use SDL1.2 in 2025 for ports of DevilutionX to older platforms that don't have accelerated graphics, which is unfortunate. It'd be nice to be able to target a single version of SDL.

FWIW, there's two separate ways of supporting accelerated palette-based output on different platforms - one is to use paletted textures (best suited for Direct3D/OpenGL/consoles) which PR #6192 implements, and the other is to use paletted display modes (best suited for GDI/DOS/RISC OS/DS), which you'd still need to use SDL 1.2 to make proper use of.

> We are waiting till it happens to migrate the project from SDL2🙂

You might as well migrate now - SDL2 doesn't officially have this functionality yet either, so you wouldn't be losing out.

**Comment by glebm** [comment]
I suppose we need palette display modes for DOS and devices like RG99, and palettes textures for PSP and such

For RG99, which is a Linux handheld, we're currently using SDL1.2.
For DOS, we're using this SDL2 port: https://github.com/jayschwa/SDL/compare/main...jayschwa:SDL:dos-vbe

**Comment by glebm** [comment]
So turns out SDL2 does actually support 8-bit display mode when using `SDL_Surface` for drawing.
It's just not exposed in `SDL_CreateWindow` but can be set immediately after with `SDL_SetWindowDisplayMode`.

I've implement it in the SDL port for DOS and it works!

**Comment by rofl0r** [comment]
> So turns out SDL2 does actually support 8-bit display mode when using SDL_Surface for drawing.
It's just not exposed in SDL_CreateWindow but can be set immediately after with SDL_SetWindowDisplayMode

is that the case for official SDL2 as well, or only for the svga fork you're using?

**Comment by glebm** [comment]
That's be up to individual video backends but at least the API supports it

**Comment by slouken** [comment]
SDL3 now supports 8-bit palette based images in the renderer API. This is in the latest main code and will be part of the 3.4.0 official release.
